### PR TITLE
Week3

### PR DESCRIPTION
**Pull request by hepnerthomas**
**Tasks:**
- [x] Write `Metals` contract
- [x] Write `ForgeMetals` contract
- [x] Deploy contracts to Goerli
- [ ] Develop frontend web application (in progress)

### CODE REVIEW

File: week3/contracts/ForgeMetals.sol [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        uint256[] memory tokenBalances = new uint256[](7);
+        tokenBalances[0] = metals.balanceOf(user, 0);
+        tokenBalances[1] = metals.balanceOf(user, 1);
+        tokenBalances[2] = metals.balanceOf(user, 2);
+        tokenBalances[3] = metals.balanceOf(user, 3);
+        tokenBalances[4] = metals.balanceOf(user, 4);
+        tokenBalances[5] = metals.balanceOf(user, 5);
+        tokenBalances[6] = metals.balanceOf(user, 6);
+        return tokenBalances;
+    }
```
Comment (dhruvinparikh) [inline week3/contracts/ForgeMetals.sol:125]: try considering Multicall - https://goerli.etherscan.io/address/0x5ba1e12693dc8f9c48aad8770482f4739beed696#code

File: week3/contracts/ForgeMetals.sol [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    function burn(address from, uint256 id, uint256 amount) external onlyOwner {
+        _burn(from, id, amount);
+    }
+
+}
+
+contract ForgeMetals {
+    Metals public metals;
+    mapping(address => uint256) public lastMintTime;
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:35]: done

File: week3/contracts/ForgeMetals.sol [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    uint256 public COOLDOWN_PERIOD_SECONDS = 0;
+
+    constructor() {
+        metals = new Metals();
+    }
+
+    function mineMetal(uint256 metalId) external {
+        require(metalId == 0 || metalId == 1 || metalId == 2, "Only base metals can be mined.");
+        require((lastMintTime[msg.sender] + COOLDOWN_PERIOD_SECONDS <= block.timestamp) || (lastMintTime[msg.sender] == 0), "Only 1 unit of a base metal can be mined per minute.");
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:44]: done

File: week3/contracts/ForgeMetals.sol [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        metals.mint(msg.sender, metalId, 1, "");
+        lastMintTime[msg.sender] = block.timestamp;
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:46]: done

File: week3/contracts/ForgeMetals.sol [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    function forgeAlloy(uint256 metalId1, uint256 metalId2) external {
+        require(metalId1 == 0 || metalId1 == 1 || metalId1 == 2, "Only base metals can be forged.");
+        require(metalId2 == 0 || metalId2 == 1 || metalId2 == 2, "Only base metals can be forged.");
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:51]: done

File: week3/contracts/ForgeMetals.sol [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        // BABBIT = TIN + ZINC
+        else if ((metalId1 == 1 && metalId2 == 2) || (metalId2 == 1 && metalId1 == 2) ) {
+            metals.mint(msg.sender, 4, 1, "");
+        }
+
+        // BRASS = COPPER + ZINC
+        else if ((metalId1 == 0 && metalId2 == 2) || (metalId2 == 0 && metalId1 == 2) ) {
+            metals.mint(msg.sender, 5, 1, "");
+        }
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:76]: done

File: week3/contracts/ForgeMetals.sol [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        uint256 metal3Amount = metals.balanceOf(msg.sender, 2);
+
+        require(metal1Amount > 0, "Not enough copper to forge gun metal.");
+        require(metal2Amount > 0, "Not enough tin to forge gun metal.");
+        require(metal3Amount > 0, "Not enough zinc to forge gun metal.");
+
+        metals.burn(msg.sender, 0, 1);
+        metals.burn(msg.sender, 1, 1);
+        metals.burn(msg.sender, 2, 1);
+        metals.mint(msg.sender, 6, 1, "");
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:92]: done

File: week3/contracts/ForgeMetals.sol [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+    function destroyAlloy(uint256 alloyId) external {
+        require(alloyId >= 3 && alloyId <= 6, "Only alloys can be destroyed.");
+        uint256 amount = metals.balanceOf(msg.sender, alloyId);
+        require(amount > 0, "Not enough alloy to destroy.");
+        metals.burn(msg.sender, alloyId, 1);
+    }
+
+    function tradeMetal(uint256 sellingMetalId, uint256 buyingMetalId) external {
+        require(buyingMetalId == 0 || buyingMetalId == 1 || buyingMetalId == 2, "Only base metals can be purchased.");
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:103]: done

File: week3/contracts/ForgeMetals.sol [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        require(sellingMetalId >= 0 && sellingMetalId <= 6, "Cannot exchange invalid metal.");
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:104]: done

File: next.config.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  env: {
+    ALCHEMY_MAINNET_URL: 'https://eth-mainnet.alchemyapi.io/v2/',
+    ALCHEMY_GOERLI_URL: 'https://eth-goerli.g.alchemy.com/v2/',
+    ALCHEMY_MAINNET_API_KEY: 'uYuc-PVHTYEIdqFyraGft8-6I-Y-45zH',
+    ALCHEMY_GOERLI_API_KEY: '8JxmEQROBcpOLz-OiqzcUyjYFnaVyLeP',
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply next.config.js:6]: done

File: week3/contracts/ForgeMetals.sol [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    function burn(address from, uint256 id, uint256 amount) external onlyOwner {
+        _burn(from, id, amount);
+    }
+
+}
+
+contract ForgeMetals {
+    Metals public metals;
+    mapping(address => uint256) public lastMintTime;
+    uint256 public COOLDOWN_PERIOD_SECONDS = 0;
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:36]: done

File: week3/contracts/ForgeMetals.sol [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+    constructor() {
+        metals = new Metals();
+    }
+
```
[the remark this answers is not in the document]
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:41]: done

File: week3/contracts/ForgeMetals.sol [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+contract ForgeMetals is Ownable {
+    Metals public metals;
+    uint256 public lastMintTime;
+    uint256 public COOLDOWN_PERIOD_SECONDS = 60;
+
+    constructor() {
+        metals = new Metals();
+        metals.transferOwnership(msg.sender);
```
Comment (dhruvinparikh) [inline week3/contracts/ForgeMetals.sol:41]: remove this line

File: week3/contracts/ForgeMetals.sol [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        transferOwnership(msg.sender);
```
Comment (dhruvinparikh) [inline week3/contracts/ForgeMetals.sol:42]: Remove this line

File: week3/contracts/ForgeMetals.sol (hunk 13 above)
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:41]: done

File: week3/contracts/ForgeMetals.sol (hunk 14 above)
Comment (hepnerthomas) [reply week3/contracts/ForgeMetals.sol:42]: done

**Pull request merged**